vscode/src/paths.ts

```diff
@@ -0,0 +1,76 @@
+import * as vscode from "vscode";
+
+export interface ExtensionPaths {
+  /** Directory with the extension's sample resources. */
+  extResources: vscode.Uri;
+
+  /** Workspace repository root. */
+  workspaceRepo: vscode.Uri;
+
+  /** Directory for analysis and resolution data files. */
+  data: vscode.Uri;
+
+  /** Directory for the extension's settings files. */
+  settings: vscode.Uri;
+
+  /** Direct path to the extension's provider settings yaml file. */
+  settingsYaml: vscode.Uri;
+
+  /** Directory to use as the working directory for the jsonrpc server. */
+  serverCwd: vscode.Uri;
+
+  /** Directory for jsonrpc server logs. */
+  serverLogs: vscode.Uri;
+}
+
+async function ensureDirectory(uri: vscode.Uri, ...parts: string[]): Promise<vscode.Uri> {
+  const joined = vscode.Uri.joinPath(uri, ...parts);
+
+  let needsCreate = true;
+  try {
+    const stat = await vscode.workspace.fs.stat(joined);
+    if (stat.type & vscode.FileType.Directory) {
+      needsCreate = false;
+    }
+  } catch {
+    needsCreate = true;
+  }
+
+  if (needsCreate) {
+    vscode.workspace.fs.createDirectory(joined);
+  }
+  return joined;
+}
+
+export async function ensurePaths(context: vscode.ExtensionContext): Promise<ExtensionPaths> {
+  const firstWorkspace = vscode.workspace.workspaceFolders?.[0];
```

Comment:
quick search for `workspace.workspaceFolders?.[0]` reveals 3 more places - it would be nice to switch them to `paths`